sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -1312,20 +1312,23 @@ std::unique_ptr<CurlNetworkConnection> CurlConnectionPool::ExtractOrCreateCurlCo
   // Libcurl setup before open connection (url, connect_only, timeout)
   if (!SetLibcurlOption(newHandle, CURLOPT_URL, request.GetUrl().GetAbsoluteUrl().data(), &result))
   {
+    curl_easy_cleanup(newHandle);
```

Comment:
As I mentioned in my other comment: Every time I see a raw handle type (see line 1303), I get worried.

The changes here to always call curl_easy_cleanup seem to be ok, but I'd rather see something like:
```c++
    class CURL_deleter {
    public:
      void operator()(CURL *handle) noexcept
      {
        if (handle != nullptr)
        {
          curl_easy_cleanup(handle);
        }
      }
    };
    using unique_CURL = std::unique_ptr<CURL, CURL_deleter>;
```

And then replace line 1303 with:
```c++
unique_CURL newHandle(curl_easy_init(), CURL_deleter());
```

The other change is that everywhere `newHandle` is referenced, it should be replaced with `newHandle.get()`.

That way, no matter whatever throws, the handle will be freed.

As the code is currently written, if you ever call a function that might throw (constructing a `std::string` for example, then you introduce a new leak. And everyone working in this codebase needs to be extra careful to ensure that they never call anything that might throw.

I'd rather the code be written from scratch to ensure it can never leak.

Reply:
If you want, I can bundle this fix as a part of the Websockets work - I already made a similar fix to WinHTTP in my websocket PR, fixing this wouldn't be out of the question.

Reply:
i think  i got it 
just pushed the change to build  and run tests